src/services/FirebaseService.ts

```diff
@@ -0,0 +1,100 @@
+import 'firebase/auth';
+
+import firebase from 'firebase/app';
+
+const config = {
+  apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
+  authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
+  projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
+  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
+  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
+  appId: process.env.REACT_APP_FIREBASE_APP_ID,
+};
+
+class Firebase {
+  private auth: firebase.auth.Auth;
+
+  constructor() {
+    if (!firebase.apps.length) {
+      firebase.initializeApp(config);
+    } else {
+      firebase.app(); // if already initialized, use that one
+    }
+    this.auth = firebase.auth();
+  }
+
+  getAuth(): firebase.auth.Auth {
+    return this.auth;
+  }
+
+  // *** Auth API ***
+  // Creates a user and signs the user in
+  doCreateUserWithEmailAndPassword(
+    email: string,
+    password: string
+  ): Promise<firebase.auth.UserCredential> {
+    return this.auth.createUserWithEmailAndPassword(email, password);
+  }
+
+  doSignInWithEmailAndPassword(
+    email: string,
+    password: string,
+    staySignedIn: boolean
+  ): Promise<firebase.auth.UserCredential> {
+    if (staySignedIn) {
+      this.auth.setPersistence(firebase.auth.Auth.Persistence.LOCAL);
+    } else {
+      this.auth.setPersistence(firebase.auth.Auth.Persistence.SESSION);
+    }
+    return this.auth.signInWithEmailAndPassword(email, password);
+  }
+
+  doSignOut = (): Promise<void> => this.auth.signOut();
+
+  doPasswordReset(email: string): void {
+    // TODO: Hide 400 error when an account with the specified email does not exist
+    this.auth.sendPasswordResetEmail(email).catch(() => {
+      // This catch block is empty because the user should not know if
+      // an account with the specified email exists
+    });
+  }
+
+  doReauthenticateUser(password: string): Promise<firebase.auth.UserCredential> {
+    const authUser = this.auth.currentUser;
+    // TODO: Redirect to 404 page if a user or user email does not exist
+    if (authUser === null) {
+      return Promise.reject(new Error('No user is logged in.'));
+    }
+    if (authUser.email === null) {
+      return Promise.reject(new Error('User email does not exist.'));
+    }
```

Comment:
Just to get my understanding better, in what situations will `authUser` and `authUser.email` be null?

Reply:
This should not ever happen in this function because it is called in the account details page where the user will be logged in and requesting to change their password or email. This is here more or less to address typescript errors and unauthorized users somehow access the account page.